src/storage-types/src/lib.rs

```diff
@@ -85,3 +85,23 @@ pub mod parameters;
 pub mod shim;
 pub mod sinks;
 pub mod sources;
+
+/// Explicitly states the contract between storage and higher levels of
+/// Materialize w/r/t which facets of objects managed by storage (e.g. sources,
+/// sinks, connections) may be altered.
+///
+/// n.b. when implementing this trait, leave a warning log with more details as
+/// to what the problem was, given that the returned error is scant on details.
```

Comment:
If tracing a warn is very important, maybe we should make it required that `fn alter_compatible(...)` returns a message? e.g.
```
fn alter_compatible(...) -> Result<(), (String, StorageError)> { ... }
```
Up to you!

Reply:
Good call––hadn't considered that patter.

Reply:
Ah, this makes nesting the logs annoying so leaving as-is but I appreciate the suggestion.